src/crimson/osd/recovery_backend.cc

```diff
@@ -225,54 +225,41 @@ RecoveryBackend::scan_for_backfill(
 {
   logger().debug("{} starting from {}", __func__, start);
   auto version_map = seastar::make_lw_shared<std::map<hobject_t, eversion_t>>();
-  return backend->list_objects(start, max).then_interruptible(
-    [this, start, version_map] (auto&& ret) {
-    auto&& [objects, next] = std::move(ret);
-    return seastar::do_with(
-      std::move(objects),
-      [this, version_map](auto &objects) {
-      return interruptor::parallel_for_each(objects,
-	[this, version_map] (const hobject_t& object)
-	-> interruptible_future<> {
-	crimson::osd::ObjectContextRef obc;
-	if (pg.is_primary()) {
-	  obc = pg.obc_registry.maybe_get_cached_obc(object);
-	}
-	if (obc) {
-	  if (obc->obs.exists) {
-	    logger().debug("scan_for_backfill found (primary): {}  {}",
-			   object, obc->obs.oi.version);
-	    version_map->emplace(object, obc->obs.oi.version);
-	  } else {
-	    // if the object does not exist here, it must have been removed
-	    // between the collection_list_partial and here.  This can happen
-	    // for the first item in the range, which is usually last_backfill.
-	  }
-	  return seastar::now();
-	} else {
-	  return backend->load_metadata(object).safe_then_interruptible(
-	    [version_map, object] (auto md) {
-	    if (md->os.exists) {
-	      logger().debug("scan_for_backfill found: {}  {}",
-			     object, md->os.oi.version);
-	      version_map->emplace(object, md->os.oi.version);
-	    }
-	    return seastar::now();
-	  }, PGBackend::load_metadata_ertr::assert_all{});
-	}
-      });
-    }).then_interruptible([version_map, start=std::move(start), next=std::move(next), this] {
-      BackfillInterval bi;
-      bi.begin = std::move(start);
-      bi.end = std::move(next);
-      bi.version = pg.get_info().last_update;
-      bi.objects = std::move(*version_map);
-      logger().debug("{} BackfillInterval filled, leaving, {}",
-                     "scan_for_backfill",
-		     bi);
-      return seastar::make_ready_future<BackfillInterval>(std::move(bi));
-    });
+  auto&& [objects, next] = co_await backend->list_objects(start, max);
+  co_await interruptor::parallel_for_each(objects, [this, version_map]
+    (const hobject_t& object) -> interruptible_future<> {
+    logger().debug("{} querying obj:{}", __func__, object);
+    auto obc_orderer = pg.obc_loader.get_obc_orderer(object);
+    auto obc_manager = pg.obc_loader.get_obc_manager(
+      obc_orderer,
+      object);
+    co_await pg.obc_loader.load_and_lock(
```

Comment:
@athanatos, out of scope - why are we still allowing users to get obcs using `ObjectConextLoader::with_obc`? Is it for legacy/future refactors reasons or am I missing something else?

Reply:
Simply to limit the scope of the previous PR -- I plan to remove users as they get refactored.